File: antia_numerical_methods/f77/sub8/bin.f
!	To calculate the modified Bessel function of first kind of
!		integral order for real argument
!
!	N : (input) Order of Bessel function required, N must be positive
!	XB : (input) Argument at which the value is required
!	BI : (output) Real array of length at least N+16+MAX(25, 5*SQRT(N))
!		which will contain the value of Bessel function of order
!		0,1,...,N. BI(I+1) will contain Bessel function of order I
!		Remaining elements of array are used as scratch space
!
!	Required routines : BI0, BI1
!
 
      SUBROUTINE BIN(N,XB,BI)
      IMPLICIT REAL*8(A-H,O-Z)
!	REPS should be less than the machine accuracy
      PARAMETER(REPS=1.D-17)
      DIMENSION BI(*)
 
51    FORMAT('  BIN FAILED AT N =',I6,'  X =',1PD13.5,'  S =',D13.5)
 
      X=ABS(XB)
      NA=ABS(N)
      IF(XB.EQ.0.0) THEN
        BI(1)=1.0
        DO 1500 I=2,NA+1
          BI(I)=0.0
1500    CONTINUE
        RETURN
      ENDIF

      IF(NA.LT.X-10.OR.NA.LT.2) THEN
!	Use the recurrence relation in forward direction
 
        BI(1)=BI0(X)
        BI(2)=BI1(X)
        DO 2000 I=2,NA
          BI(I+1)=-2*(I-1)*BI(I)/X+BI(I-1)
2000    CONTINUE

      ELSE IF(X.LE.4.0) THEN

!	Use series expansion to calculate BI(NA), BI(NA-1)
        XA=X*X/4.0
        T0=X/2.0
        DO 2400 I=2,NA
          T0=T0*X/(2.0*I)
          IF(I.GE.NA-1) THEN
            T=T0
            S=T0
            DO 2200 J=1,50
              T=T*XA/(J*(J+I))
              S=S+T
              IF(ABS(T).LT.ABS(S*REPS)) GO TO 2300
2200        CONTINUE
2300        BI(I+1)=S
          ENDIF
2400    CONTINUE

        DO 2600 I=NA-1,1,-1
          BI(I)=2.*I*BI(I+1)/X+BI(I+2)
2600    CONTINUE
      ELSE
 
!	Use the recurrence relation in backward direction
        N1=NA+MAX(25.D0,5.*SQRT(1.D0*NA))
        IF(X.LT.NA/6.0) N1=NA+MAX(25.D0,5.*SQRT(1.D0*NA))/LOG(NA*0.5/X)
        IF(NA.LT.X+15) N1=N1+15
        IF(N1-NA.LT.15) N1=15+NA
        BI(N1+1)=0.0
        BI(N1)=1.0
        DO 3200 I=N1-1,1,-1
          BI(I)=2.*I*BI(I+1)/X+BI(I+2)
3200    CONTINUE

        S=BI(1)/BI0(X)
        DO 3600 I=1,NA+1
          BI(I)=BI(I)/S
3600    CONTINUE
!	If ABS(BI(NA+2))<1/REPS, then the required accuracy may not be achieved
!	hence printout an error message
        IF(ABS(BI(NA+2))*REPS.LT.1.0) PRINT 51,N,X,BI(NA+2)
      ENDIF
 
      IF(XB.LT.0.0) THEN
        DO 3800 I=2,NA+1,2
          BI(I)=-BI(I)
3800    CONTINUE
      ENDIF
      END

File: projects/antia_numerical_methods@Beliavsky/c/fun/bin.c
/*	To calculate the modified Bessel function of first kind of
		integral order for real argument

	N : (input) Order of Bessel function required, N must be positive
	XB : (input) Argument at which the value is required
	BI : (output) Array of length at least N+16+MAX(25, 5*sqrt(N))
		which will contain the value of Bessel function of order
		0,1,...,n. BI[i] will contain Bessel function of order i
		Remaining elements of array are used as scratch space

	Required functions : BI0, BI1
*/
#include <stdio.h>
#include <math.h>

double bi0(double x);
double bi1(double x);

void bin(int n, double xb, double bi[])

{
	int i,j,n1,na;
/*	REPS should be less than the machine accuracy */
	double x,s,t,t0,xa, reps=1.e-17;

	x=fabs(xb);
	na=abs(n);
	if(xb==0.0) {
		bi[0]=1.0;
		for(i=1; i<=na; ++i) bi[i]=0.0;
		return;
	}

	else if(na<x-10 || na<2) {
/*	Use the recurrence relation in the forward direction  */
		bi[0]=bi0(x);
		bi[1]=bi1(x);
		for(i=2; i<=na; ++i) bi[i]=-2*(i-1)*bi[i-1]/x+bi[i-2];
	}

	else if(x<=4.0) {
/*	Use series expansion to calculate  bi[na], bi[na-1] */
		xa=x*x/4.0;
		t0=x/2.0;
		for(i=2; i<=na; ++i) {
			t0=t0*x/(2.0*i);
			if(i>=na-1) {
				t=t0; s=t0;
				for(j=1; j<51; ++j) {
					t=t*xa/(j*(j+i));
					s=s+t;
					if(fabs(t)<fabs(s*reps)) break;
				}
				bi[i]=s;
			}
		}
		for(i=na-1; i>=1; --i) bi[i-1]=2.*i*bi[i]/x+bi[i+1];
	}


	else {
/*	Use the recurrence relation in the backward direction  */
		n1=5.*sqrt((double) na); if(n1<25) n1=25;
		if(x<na/6.0) n1=n1/log(na*0.5/x);
		n1=n1+na;
		if(na<x+15.0) n1=n1+15;
		if(n1-na<15) n1=na+15;
		bi[n1]=0.0; bi[n1-1]=1.0;
		for(i=n1-1; i>=1; --i) bi[i-1]=2.*i*bi[i]/x+bi[i+1];

		s=bi[0]/bi0(x);
		for(i=0; i<=na; ++i) bi[i]=bi[i]/s;
/*	If fabs(bi[na+1])<1/reps, then the required accuracy may not be achieved
	hence printout an error message */
		if(fabs(bi[na+1]*reps)< 1.0) printf(" bin failed at n = %d  x = %e  s = %e \n", n,x,bi[na+1]);
	}

	if(xb<0.0) {
		for(i=1; i<=na; i += 2) bi[i]= -bi[i];
	}
	return;
}

 
